### PR TITLE
refactor and separation

### LINKED ISSUE #16: bad local
The `nthr`variable here; https://github.com/openresty/lua-resty-upstream-healthcheck/blob/master/lib/resty/upstream/healthcheck.lua#L322 

Should not be a local. It shadows the outer scope. as such the check slightly further down, where we wait for all spawned threads to finish will never run, because the outer scoped `nthr` will still be `nil`.

### LINKED ISSUE #17: bad id reference
see https://github.com/openresty/lua-resty-upstream-healthcheck/blob/master/lib/resty/upstream/healthcheck.lua#L282

This line assumes that `iterator - 1` equals the `id` of the peer. That is a bad assumption. It should use `peers[iterator].id` instead probably to use the true `id` value provided by the upstream module.

The code has multiple occurrences of this issue.

### PR DESCRIPTION
**Pull request by Tieske**
Major refactor
(branch name loadbalancer is a bad name, but this is a first step to creating it)

Closes #12 
Fixes #16 
Fixes #17 

I started of by extracting the 'upstream manager' but then ran into the issue of many synchronizations due to more dynamic upstreams (adding/removing peers, see also #17). So I removed the internal event-like mechanism for the updates and rewrote it in a proper library (see https://github.com/Mashape/lua-resty-worker-events/blob/master/README.markdown).
The effect is that status changes propagate much faster, as there is no wait for the second worker to expire its timer.

The 'upstream manager' is configurable (to replace it when using proxypass directive instead of upstreams for example), but defaults to the resty upstream module. It should implement;
- `get_upstreams()`
- `get_peers(upstream)`

If it needs to act upon status changes, it can subscribe to `peer_status` events. See the [upstream module wrapper](https://github.com/Tieske/lua-resty-upstream-healthcheck/blob/loadbalancer/lib/resty/upstream/healthcheck/hcu_wrapper.lua) for an example.

It cleanly separates the upstream stuff from the healthchecker, and offloads all synchronization issues to the new event module. 

It's up to date with master and all tests pass.

let me know what you think.

### CODE REVIEW

**Comment by agentzh** [comment]
@Tieske Thanks for the hard work on this!

@doujiang24 Will you have a look at this PR? Thanks!

**Comment by evanccnyc** [comment]
Any chance this patch can be looked at anytime soon?

File: lib/resty/upstream/healthcheck.lua [hunk 1]
```diff
@@ -4,23 +4,41 @@ local ERR = ngx.ERR
...
 local ipairs = ipairs
 local ceil = math.ceil
 local spawn = ngx.thread.spawn
 local wait = ngx.thread.wait
 local pcall = pcall
+local cjson = require("cjson.safe").new()
+
+-- upstream keys (both get upstream name appended)
+local KEY_LOCK = "resty-healthcheck:l:" -- lock
+local KEY_DATA = "resty-healthcheck:d:" -- serialized healthcheck json data
```
Comment (agentzh) [inline lib/resty/upstream/healthcheck.lua]: The key prefix is too long and is a waste of shm storage I'm afraid.

File: lib/resty/upstream/healthcheck.lua [hunk 2]
```diff
@@ -4,23 +4,41 @@ local ERR = ngx.ERR
...
 local concat = table.concat
+local insert = table.insert
 local tonumber = tonumber
 local tostring = tostring
 local ipairs = ipairs
 local ceil = math.ceil
 local spawn = ngx.thread.spawn
 local wait = ngx.thread.wait
 local pcall = pcall
+local cjson = require("cjson.safe").new()
```
Comment (agentzh) [inline lib/resty/upstream/healthcheck.lua]: I'm not feeling very comfortable with depending on lua-cjson.

File: lib/resty/upstream/healthcheck.lua [hunk 3]
```diff
@@ -66,162 +72,101 @@ local function debug(...)
...
+    debug(event,", ", data.name)
+    return events.post(_M.events._source, event, data)
+end
+
+-- registers a single healthcheck failure for a peer, marks peer as
+-- down if the threshold of failures is exceeded
+local function peer_fail(checker, peer)
+    debug("peer ", peer.name, " was checked to be not ok")
+
+    peer.fails = math.min((peer.fails or 0) + 1, checker.fall)
```
Comment (agentzh) [inline lib/resty/upstream/healthcheck.lua]: We should avoid the table lookup in `math.min`, I think?

File: lua-resty-upstream-healthcheck-0.1.0-1.rockspec [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+}
+dependencies = {
+}
+build = {
+   type = "builtin",
+   modules = { 
+     ["resty.upstream.healthcheck"] = "lib/resty/upstream/healthcheck.lua",
+     ["resty.upstream.healthcheck.hcu_wrapper"] = "lib/resty/upstream/healthcheck/hcu_wrapper.lua",
+   }
+}
```
Comment (agentzh) [inline lua-resty-upstream-healthcheck-0.1.0-1.rockspec]: Sorry, I don't want to include rockspec.

**Comment by agentzh** [comment]
I think I like this library to be self-contained. Now this PR adds 2 external Lua library dependencies which I don't quite like. Also the code base is now much more complicated while I'd like to keep things simple. Maybe we should do a separate lua-resty-healtcheck library for the context of balancer_by_lua* and etc?

**Comment by Tieske** [comment]
since we've taken another approach and no longer use the upstreams from the config, I'm closing this.

**Pull request closed without merging**